### ISSUE TITLE
Deprecate `recompute` on `Helper`?

### ISSUE THREAD

**Issue by chancancode** [emberjs/rfcs#633]
The point of the `recompute` method is that you call it to trigger Ember to re-render the helper in some kind of callback, such as observers, promise callbacks, etc.

Since helpers are now auto-tracked, it should be possible to refactor any usage of `recompute` into setting tracked state and accomplish the same thing. In most cases, the transformation is pretty straightforward.

Before:

```js
class MyHelper extends Helper {
  @service foo;

  @observes('foo.bar.[]') onBarChanged() {
    this.recompute();
  }

  compute() {
    return this.foo.bar.map(...);
  }
}
```

After:

```js
class MyHelper extends Helper {
  @service foo;

  @tracked bar = this.foo.bar;

  @observes('foo.bar.[]') onBarChanged() {
    this.bar = this.foo.bar;
  }

  compute() {
    return this.bar.map(...);
  }
}
```

The example uses an observer, because that is a common kind of callback used in this position, but it's ultimately unrelated to my point.I also think there are better ways to accomplish the same thing without using observers, in general, but that is also besides the point here.

For prosperity, here is a second example that doesn't use observers:

Before:

```js
const UNINITIALIZED = Object.freeze({});

class AwaitHelper extends Helper {
  resolved = undefined;
  promise = UNINITIALIZED;

  compute([promise]) {
    if (promise !== this.promise) {
      this.resolved = undefined;
      this.promise = promise;

      Promise.resolve(resolved => {
        if (promise === this.promise) {
          this.resolved = resolved;
          this.recompute();
        }
      });
    }

    return this.resolved;
  }

  willDestroy() {
    this.promise = UNINITIALIZED;
  }
}
```

After:

```js
const UNINITIALIZED = Object.freeze({});

class AwaitHelper extends Helper {
  @tracked resolved = undefined;
  promise = UNINITIALIZED;

  compute([promise]) {
    if (promise !== this.promise) {
      this.resolved = undefined;
      this.promise = promise;

      Promise.resolve(resolved => {
        if (promise === this.promise) {
          this.resolved = resolved;
        }
      });
    }

    return this.resolved;
  }

  willDestroy() {
    this.promise = UNINITIALIZED;
  }
}
```
 
The general pattern is: instead of calling `recompute()`, just refactor into setting a piece of tracked state and use that tracked state in the `compute()` method.

Would be interested to see if there are any valid use cases this doesn't cover.

**Comment by rwjblue** [comment]
I think it is a _tad_ difficult to suggest "just use `@tracked`" at the moment, since we _also_ expect addons to continue supporting ember-source@3.12 which doesn't support tracking concepts. It is also non-trivial to write code that works for both styles (without completely duplicating the helper).

Either way, I agree with this direction. My point above is that we should _probably_ tie this deprecation to land no earlier than 3.21, at which point 3.12 will no longer be an active concern (since 3.16 and 3.20 would be the active LTS versions). Note: 3.21 is current ember-source master, so this is still a great time to push this forward. :smile:

**Comment by chancancode** [comment]
> 3.21 is current ember-source master

What a time to be alive 😄

**Comment by mixonic** [comment]
We discussed this again today at the framework meeting. As the tracked system is now very, very common, this proposal is definitely ready for RFC and advancement.

**Comment by wagenet** [comment]
@mixonic who will write the RFC?

**Comment by wagenet** [comment]
Should this be considered with #832 or separately?

**Comment by bertdeblock** [comment]
Is this still something we want to do as is? If so, I could try and write the RFC.

**Comment by ijlee2** [comment]
Before an author pursues deprecating `recompute` via an RFC, I think they should provide concrete examples of how `ember-intl` (i.e. frequently used addons) can move away from it.

Since `5.x` (even earlier), `ember-intl` has been [using `recompute` to update translations and formatted items](https://github.com/ember-intl/ember-intl/blob/main/packages/ember-intl/addon/helpers/t.ts#L26), because `formatjs` (a 3rd-party library) is responsible for locales and any changes to them. At the moment, I'm not sure if there's a way to refactor `ember-intl`'s `t` service so that the helpers won't need `recompute`.

**Comment by bertdeblock** [comment]
That event-based approach was introduced before `@tracked` was a thing. `_locale` on the `intl` service is now `@tracked`, so you just have to consume it. The tests even pass without `this.intl.onLocaleChanged(this.recompute, this);`.

**Comment by ijlee2** [comment]
@bertdeblock Indeed, thanks for checking! I verified through CI that it should be okay to remove `recompute` in `ember-intl`.

https://github.com/ember-intl/ember-intl/pull/1932/checks
